File: python/src/Sponser.py
from dataclasses import dataclass, asdict
import json
from .User import fullName, User

@dataclass
class Sponser:
    name: fullName
    message: str

    @staticmethod
    def new_sponsor():
        print("Have your name and a message of your choise be put on a brass plaque, and also support as with a $200 fee <3")
        first_name = User.get_input("What is your first name?", lambda s: len(s) > 1)
        last_name = User.get_input("What is your last name?", lambda s: len(s) > 1)
        message = User.get_input("What would you like your plank to say? ", lambda s: len(s) <= 128)

        fullname: fullName = fullName(first_name, last_name)

        sponser = Sponser.check_sponsor_dep(Sponser(fullname, message))

        print("Are you sure you want to continue with this info?")
        print(sponser)

        continue_sponser = User.get_input("Continue? (yes/no)", User.validate_bool).lower() in "yes"

        if continue_sponser:
            return sponser

    @staticmethod
    def check_sponsor_dep(sponser: 'Sponser'):
        sponsers = []
        with open("./sponsers.json", "r") as json_lines:
            content = json_lines.readlines()
            for i in content:
                json_object = json.loads(i)
                back_from_json: Sponser = Sponser(**json_object)

                # Make sure the name is formated correctly in the object
                fullname: fullName = fullName(**json_object["name"])
                back_from_json.name = fullname
                sponsers.append(back_from_json)

            for i in sponsers:
                if i.name == sponser.name:
                    print("You're already sponsoring")
                    continue_sponser = User.get_input("Continue as someone else? (yes/no)", User.validate_bool) in "yes"
                    if continue_sponser:
                        return Sponser.new_sponsor()
                    else:
                        exit(0x45)
            return sponser

    @staticmethod
    def save_sponsor(sponser):
        json_string = json.dumps(asdict(sponser))

        print(json_string, file=open("./sponsers.json", "a"))

    @staticmethod
    def get_sponsers():
        sponsers = []
        with open("./sponsers.json", "r") as json_lines:
            content = json_lines.readlines()
            for i in content:
                json_object = json.loads(i)
                back_from_json: Sponser = Sponser(**json_object)

                # Make sure the name is formated correctly in the object
                fullname: fullName = fullName(**json_object["name"])
                back_from_json.name = fullname

                name = f"name: {back_from_json.name.first_name} {back_from_json.name.last_name}\n"
                message = f"message: {back_from_json.message}\n"

                stringToReturn = name + message

                sponsers.append(stringToReturn)

            for i in sponsers:
                print(i)

File: go/pkg/sponsor/sponser.go
package sponsor

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"os"

	"github.com/yonikosiner/python-deez-nuts-with-sea-men/pkg/user"
	"github.com/yonikosiner/python-deez-nuts-with-sea-men/utils"
)

func (s *Sponser) NewSponser() Sponser {
	fmt.Println("Have your name and a message of your choise be put on a brass plaque, and also support as with a $200 fee <3")

	firstName := utils.AskForInput("Enter your first name? ")
	lastName := utils.AskForInput("Enter your last name? ")
	fullname := user.FullName{firstName, lastName}

	var message string

	message = utils.AskForInput("What would you like your plank to say?")

	if len(message) > 64 {
		newMessage := utils.StringToBoolean(utils.AskForInput("Your message is to long, would you like to enter a new message? (yes/no) "))
		if newMessage {
			message = utils.AskForInput("Entere in your new message ")
		}
	}

	sponser := s.checkSponsorDup(Sponser{fullname, message})

	return sponser
}

func (s *Sponser) checkSponsorDup(sponser Sponser) Sponser {
	sponsers := s.getAllSponsors()

	for _, value := range sponsers {
		if sponser.Name == value.Name {
			log.Fatal("You're already sponsoring")
			contiuneSponser := utils.StringToBoolean(utils.AskForInput("Continue as someone else? (yes/no) "))

			if contiuneSponser {
				return s.NewSponser()
			}
		}
	}
	return sponser
}

func (s *Sponser) SaveSponsor(sponser Sponser) {
	b, err := json.Marshal(sponser)

	if err != nil {
		log.Fatal(err)
	}

	f, err := os.OpenFile("./sponsers.json", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)

	if err != nil {
		log.Fatal(err)
	}

	if _, err := f.Write([]byte(fmt.Sprintf("%s\n", b))); err != nil {
		log.Fatal(err)
	}
	if err := f.Close(); err != nil {
		log.Fatal(err)
	}
}

func (s *Sponser) GetSponers() []string {
	sponsers := s.getAllSponsors()

	var sponsersReturn []string

	for _, value := range sponsers {
		name := fmt.Sprintf("Name: %s %s\n", value.Name.FirstName, value.Name.LastName)
		message := fmt.Sprintf("Message: %s", value.Message)

		stringReturn := name + message
		sponsersReturn = append(sponsersReturn, stringReturn)
	}

	return sponsersReturn
}

func (s *Sponser) getAllSponsors() []Sponser {
	b, err := os.Open("./sponsers.json")

	if err != nil {
		log.Fatal(err)
	}

	scanner := bufio.NewScanner(b)
	scanner.Split(bufio.ScanLines)

	var text []string

	for scanner.Scan() {
		text = append(text, scanner.Text())
	}

	b.Close()

	var sponsers []Sponser

	for _, value := range text {
		var data Sponser
		err := json.Unmarshal([]byte(value), &data)

		if err != nil {
			log.Fatal(err)
		}

		sponsers = append(sponsers, data)
	}

	return sponsers
}
